File: Aula 07/exemplos_agregacao_composicao.js
class Usuario {
  constructor(nome, senha) {
    this.nome = nome;
    this.senha = senha;
  }
}

class Sistema {
  constructor() {
    this.usuarios = [];
  }

  adicionarUsuario(usuario) {
    this.usuarios.push(usuario);
  }

  autenticar(nome, senha) {
    const usuario = this.usuarios.find(
      (item) => item.nome === nome && item.senha === senha
    );

    return usuario
      ? `Usuario ${nome} autenticado com sucesso.`
      : "Falha na autenticacao.";
  }
}

class Token {
  constructor(valor = "XYZ789") {
    this.valor = valor;
  }
}

class Autenticacao {
  constructor() {
    // Composição: o token é criado dentro da autenticacao
    this.token = new Token();
  }

  exibirToken() {
    return `Token gerado: ${this.token.valor}`;
  }
}

class Motorista {
  constructor(nome) {
    this.nome = nome;
  }
}

class Motor {
  constructor(potencia) {
    this.potencia = potencia;
  }
}

class Carro {
  constructor(modelo, potenciaMotor) {
    this.modelo = modelo;
    this.motorista = null;
    // Composição: o carro cria seu proprio motor
    this.motor = new Motor(potenciaMotor);
  }

  definirMotorista(motorista) {
    // Agregação: o motorista ja existia antes da associacao
    this.motorista = motorista;
  }

  mostrarDados() {
    const nomeMotorista = this.motorista ? this.motorista.nome : "Sem motorista";
    return `Carro: ${this.modelo} | Motor: ${this.motor.potencia} cv | Motorista: ${nomeMotorista}`;
  }
}

class Cliente {
  constructor(nome) {
    this.nome = nome;
  }
}

class Historico {
  constructor() {
    this.movimentos = [];
  }

  adicionarMovimento(descricao) {
    this.movimentos.push(descricao);
  }
}

class Conta {
  constructor(numero, titular) {
    this.numero = numero;
    this.titular = titular;
    // Composição: o historico pertence a conta
    this.historico = new Historico();
  }

  depositar(valor) {
    this.historico.adicionarMovimento(`Deposito de R$ ${valor}`);
  }
}

class Banco {
  constructor(nome) {
    this.nome = nome;
    this.clientes = [];
  }

  adicionarCliente(cliente) {
    // Agregação: o cliente existe independentemente do banco
    this.clientes.push(cliente);
  }
}

console.log("=== AUTENTICACAO ===");
const usuario = new Usuario("Ana", "1234");
const sistema = new Sistema();
sistema.adicionarUsuario(usuario);
console.log(sistema.autenticar("Ana", "1234"));

const auth = new Autenticacao();
console.log(auth.exibirToken());

console.log("\n=== CARRO ===");
const motorista = new Motorista("Carlos");
const carro = new Carro("Fusca", 85);
carro.definirMotorista(motorista);
console.log(carro.mostrarDados());

console.log("\n=== BANCO ===");
const cliente = new Cliente("Marina");
const banco = new Banco("Banco POO");
banco.adicionarCliente(cliente);

const conta = new Conta("0001", cliente.nome);
conta.depositar(500);

console.log(`Banco: ${banco.nome}`);
console.log(`Cliente agregado: ${banco.clientes[0].nome}`);
console.log(`Conta: ${conta.numero} | Titular: ${conta.titular}`);
console.log("Historico da conta:", conta.historico.movimentos);

File: Aula 07/exemplos_agregacao_composicao.py
class Usuario:
    def __init__(self, nome, senha):
        self.nome = nome
        self.senha = senha


class Sistema:
    def __init__(self):
        self.usuarios = []

    def adicionar_usuario(self, usuario):
        self.usuarios.append(usuario)

    def autenticar(self, nome, senha):
        for usuario in self.usuarios:
            if usuario.nome == nome and usuario.senha == senha:
                return f"Usuario {nome} autenticado com sucesso."
        return "Falha na autenticacao."


class Token:
    def __init__(self, valor="ABC123"):
        self.valor = valor


class Autenticacao:
    def __init__(self):
        # Composição: o token nasce dentro da autenticacao
        self.token = Token()

    def exibir_token(self):
        return f"Token gerado: {self.token.valor}"


class Motorista:
    def __init__(self, nome):
        self.nome = nome


class Motor:
    def __init__(self, potencia):
        self.potencia = potencia


class Carro:
    def __init__(self, modelo, potencia_motor):
        self.modelo = modelo
        self.motorista = None
        # Composição: o carro cria seu proprio motor
        self.motor = Motor(potencia_motor)

    def definir_motorista(self, motorista):
        # Agregação: o motorista ja existe e eh associado ao carro
        self.motorista = motorista

    def mostrar_dados(self):
        nome_motorista = self.motorista.nome if self.motorista else "Sem motorista"
        return (
            f"Carro: {self.modelo} | "
            f"Motor: {self.motor.potencia} cv | "
            f"Motorista: {nome_motorista}"
        )


class Cliente:
    def __init__(self, nome):
        self.nome = nome


class Historico:
    def __init__(self):
        self.movimentos = []

    def adicionar_movimento(self, descricao):
        self.movimentos.append(descricao)


class Conta:
    def __init__(self, numero, titular):
        self.numero = numero
        self.titular = titular
        # Composição: o historico faz parte da conta
        self.historico = Historico()

    def depositar(self, valor):
        self.historico.adicionar_movimento(f"Deposito de R$ {valor}")


class Banco:
    def __init__(self, nome):
        self.nome = nome
        self.clientes = []

    def adicionar_cliente(self, cliente):
        # Agregação: o cliente pode existir fora do banco
        self.clientes.append(cliente)


if __name__ == "__main__":
    print("=== AUTENTICACAO ===")
    usuario = Usuario("Ana", "1234")
    sistema = Sistema()
    sistema.adicionar_usuario(usuario)
    print(sistema.autenticar("Ana", "1234"))

    auth = Autenticacao()
    print(auth.exibir_token())

    print("\n=== CARRO ===")
    motorista = Motorista("Carlos")
    carro = Carro("Fusca", 85)
    carro.definir_motorista(motorista)
    print(carro.mostrar_dados())

    print("\n=== BANCO ===")
    cliente = Cliente("Marina")
    banco = Banco("Banco POO")
    banco.adicionar_cliente(cliente)

    conta = Conta("0001", cliente.nome)
    conta.depositar(500)

    print(f"Banco: {banco.nome}")
    print(f"Cliente agregado: {banco.clientes[0].nome}")
    print(f"Conta: {conta.numero} | Titular: {conta.titular}")
    print("Historico da conta:", conta.historico.movimentos)
